tools/generate-opam-sources.sh

```diff
@@ -0,0 +1,80 @@
+#! /usr/bin/env bash
+# Emit a table of opam file and where to download it. This can be used
+# to update Opam files to the version in a code repository
+
+EXTRA="packages/xs-extra"
+XAPI="xapi-project"
+XSER="xenserver"
+
+MAP="\
+nbd-tool                  $XAPI/nbd
+rrd2csv                   $XSER/rrd2csv
+xen-api-client-lwt        $XAPI/xen-api-client
+xen-api-client            $XAPI/xen-api-client
+xen-api-client-async      $XAPI/xen-api-client
+xapi-database             $XAPI/xen-api
+xapi-client               $XAPI/xen-api
+xapi-consts               $XAPI/xen-api
+xapi-types                $XAPI/xen-api
+xapi                      $XAPI/xen-api
+xapi-cli-protocol         $XAPI/xen-api
+xapi-datamodel            $XAPI/xen-api
+xe                        $XAPI/xen-api
+xapi-xenopsd-xc           $XAPI/xenopsd
+xapi-xenopsd              $XAPI/xenopsd
+xapi-xenopsd-simulator    $XAPI/xenopsd
+xapi-squeezed             $XAPI/squeezed
+xapi-networkd             $XAPI/xcp-networkd
+forkexec                  $XAPI/forkexecd
+xapi-forkexecd            $XAPI/forkexecd
+message-switch-lwt        $XAPI/message-switch
+message-switch            $XAPI/message-switch
+message-switch-async      $XAPI/message-switch
+message-switch-cli        $XAPI/message-switch
+message-switch-core       $XAPI/message-switch
+message-switch-unix       $XAPI/message-switch
+xapi-storage              $XAPI/xapi-storage
+vhd-tool                  $XAPI/vhd-tool
+xen-api-sdk               $XAPI/xen-api-sdk
+xapi-plugin               $XAPI/ocaml-xapi-plugin
+xapi-storage-script       $XAPI/xapi-storage-script
+xapi-storage-cli          $XAPI/sm-cli
+xapi-nbd                  $XAPI/xapi-nbd
+vncproxy                  $XAPI/vncproxy
+xapi-libs-transitional    $XAPI/xen-api-libs-transitional
+xenctrlext                $XAPI/xen-api-libs-transitional
+uuid                      $XAPI/xen-api-libs-transitional
+http-svr                  $XAPI/xen-api-libs-transitional
+sexpr                     $XAPI/xen-api-libs-transitional
+pciutil                   $XAPI/xen-api-libs-transitional
+stunnel                   $XAPI/xen-api-libs-transitional
+gzip                      $XAPI/xen-api-libs-transitional
+sha1                      $XAPI/xen-api-libs-transitional
+xml-light2                $XAPI/xen-api-libs-transitional
+zstd                      $XAPI/xen-api-libs-transitional
+rrdd-plugin               $XAPI/ocaml-rrdd-plugin
+xapi-rrdd-plugin          $XAPI/ocaml-rrdd-plugin
+xapi-rrd-transport-utils  $XAPI/rrd-transport
+xapi-rrd-transport        $XAPI/rrd-transport
+xapi-rrdd                 $XAPI/xcp-rrdd
+rrd-transport             $XAPI/rrd-transport
+xenops                    $XAPI/xenops
+xapi-xenops               $XAPI/xenops
+xapi-xenops-utils         $XAPI/xenops
+wsproxy                   $XAPI/wsproxy
+xapi-idl                  $XAPI/xcp-idl
+xapi-tapctl               $XAPI/tapctl"
```

Comment:
Can you use `opam info -f url.src: ${package}` to get the URLs directly from opam? e.g. `opam info -f url.src: xapi-datamodel `gives `"https://github.com/xapi-project/xen-api/archive/master/master.tar.gz"`

Reply:
Hmm, that would need to get opam to be configured properly. The way the script is set up makes it independent from opam.

In any case we still need that url chunk to curl the opam file from raw.githubusercontent.com

Reply:
If the script can run without opam that is a useful property to preserve.
What if you fetch the github URL from the existing opam file instead? i.e. grep for either the dev-repo or the source url and use sed to find the parts you need and then put it under raw.githubusercontent. Then we don't need to maintain this list of mappings (in case we get more subpackages in the future, in particular this doesn't include the varstored-guard opam file for example)

Reply:
There is an ocaml tool to extract fields from Opam files: 

* https://opam.ocaml.org/packages/opam-ed/

Maybe we should add it to xs-extra such that we have it available?

Reply:
I don't like maintaining this either.

I can try using a list of "watched" repositories, the script would download all the opam files it finds on those, create directories and files and add the `url.src` to them as needed.

This would mean the source of how xs-extra is populated remains inside xs-opam, while on the repos we can maintain the files as needed.

Ideally I'd also like some kind of CI that auto-updates xs-opam when detecting a difference after running this script.

Reply:
I think we can leave this aspect of the PR as-is and open another PR to look for missing opam files in watched repositories.